features/caritas-validate-preferences-paper-application-#159245564

fddayan opened the pull request:
https://www.pivotaltracker.com/story/show/159245564

fddayan committed:
[#159245564] Adding preference validations WIP

fddayan committed:
[#159245564] Adding validations for the rest preferences

fddayan committed:
[#159246205] Cleanup

fddayan committed:
[#159246205] Removed file not used

akegan commented on app/javascript/components/applications/application_form/PrimaryApplicantSection.js:
```diff
@@ -16,35 +16,10 @@ let mailingAddressFieldMap = {
-  return years >= 18
-}
-
-const isValidDate = (dateOfBirth) => {
```
looks like this one didn't make it over to your validation page.

fddayan replied:
fixed

akegan commented on app/javascript/components/applications/application_form/PrimaryApplicantSection.js:
```diff
@@ -16,35 +16,10 @@ let mailingAddressFieldMap = {
-    return null
-}
+const validateError = validate({
+  date_of_birth: validate.isOldEnough("The primary applicant must be 18 years of age or older"),
```
missing isValidDate here

chrisdolendo replied:
I don't think `date_of_birth` and `first_name` is relevant in this version of the code no? I think this should be `DOB` and `firstName`.

akegan commented on app/javascript/components/applications/application_form/PrimaryApplicantSection.js:
```diff
@@ -102,22 +68,20 @@ const PrimaryApplicantSection = ({formApi, editValues }) => {
+                <Field formApi={formApi} field='DOB' label='DOB' blockNote='- YYYY-MM-DD (required)' >
+                  {(field, classNames) => (
+                    <Text
+                      field="DOB"
```
DOB or date_of_birth? it's date_of_birth on line 20 above

akegan commented on app/javascript/components/applications/application_form/PrimaryApplicantSection.js:
```diff
@@ -102,22 +68,20 @@ const PrimaryApplicantSection = ({formApi, editValues }) => {
-                    className={classNames(errorClassName(formApi, 'DOB'))} />
-                  <FormError formApi={formApi} field='DOB'/>
-                </div>
+                <Field formApi={formApi} field='DOB' label='DOB' blockNote='- YYYY-MM-DD (required)' >
```
this block note is actually incorrect for the way the date is input in chrome with the HTML5 date field:

![image](https://user-images.githubusercontent.com/11825994/44112809-ff780bec-9fba-11e8-9f16-f679f4ceb1ba.png)

chrisdolendo replied:
The label on this is old... it should be `Date of Birth (required)`

akegan reviewed (dismissed):
Just a start. A couple of questions, but most importantly, DOB validation is broken on this PR. We're no longer getting inline errors at all for <18years old or invalid date, either from tabbing away or on save

fddayan committed:
[#159245564] Adding isValidDate and cleanups

fddayan committed:
[#159245564] Fix DOB label

akegan commented:
Date of birth seems good now, but multiple preferences are still wonky.

I tried COP and didn't get any red highlighting until I hit submit, but adding other preferences leads to highlighting before clicking into the fields at all

akegan commented:
![issuewithvalidatepreferences](https://user-images.githubusercontent.com/11825994/44163941-9a98eb00-a079-11e8-90ff-b4bd44635a08.gif)

fddayan committed:
[#159245564] Validate fields on change

fddayan committed:
[#159245564] Snapshots updated

fddayan commented:
@akegan as we talked over slack: that problem in the gif image is out of the scope of this story. It's not part of the AC and @slowbot  will have to confirm if that is the desire behavior. It's better if we capture that work in a different story

akegan reviewed (approved):
LGTM now, should we be wrapping every field in a "Field" component going forward?

chrisdolendo reviewed (approved):
:+1: Looks good to me!

callaghanc commented on app/javascript/utils/reactFormUtils.js:
```diff
@@ -0,0 +1,89 @@
+// Shamles copy paste from react-form/redux/utils, Fed
```
@fddayan Seems like lodash already offers most of these: https://lodash.com/docs/4.17.10. Can you replace any usage of these with lodash methods where possible, keeping just the ones that are not offered by lodash? These are totally fine, but it's less complicated to stick to one set of util functions as much as possible.

fddayan replied:
I tried replacing the lodash methods `isArray`, `isObject`, `flattenDeep` and validations stopped working. There might be subtle differences.I did not dig into why.

fddayan replied:
@callaghanc as we talked on slack. I added a NOTE and TODO comment.

callaghanc replied:
🙏

callaghanc committed:
[#159245564] Adds labels to ADHP and custom pref dropdowns

Also does some layout adjustments in the paper application form Preferences section.

fddayan committed:
[#159245564] Note added

akegan commented:
looks like the PR @chrisdolendo just merged caused merge conflicts on this one

akegan committed:
Merge branch 'caritas-feedback-release' into features/caritas-validate-preferences-paper-application-#159245564
